### PR TITLE
Resample var

### PR DESCRIPTION
**Pull request by roflmaostc**

### CODE REVIEW

File: src/fftshift_alternatives.jl [hunk 1]
```diff
@@ -55,26 +55,26 @@ end
...
 function ifftshift2d(mat::AbstractArray{T, N}) where {T, N}
     ifftshift(mat,(1,2))
 end
 
 """
     fftshift2d_view(mat::AbstractArray{T, N}) where {T, N}
-    Short-hand for fftshift_view(mat, (1,2)) performing only a 2D ft.
-    See fft for details.
+    Short-hand for fftshift_view(mat, (1,2)).
+    See fftshift_view for details.
```
Comment (roflmaostc) [inline src/fftshift_alternatives.jl]: `fftshift_view`

File: src/fftshift_alternatives.jl [hunk 2]
```diff
@@ -55,26 +55,26 @@ end
...
 function fftshift2d(mat::AbstractArray{T, N}) where {T, N}
     fftshift(mat,(1,2))
 end
 
 """
     ifftshift2d(mat::AbstractArray{T, N}) where {T, N}
-    Short-hand for ifftshift(mat, (1,2)) performing only a 2D inverse ft.
-    See ifft for details.
+    Short-hand for ifftshift(mat, (1,2)).
+    See ifftshift for details.
```
Comment (roflmaostc) [inline src/fftshift_alternatives.jl]: `ifftshift`

File: src/fftshift_alternatives.jl [hunk 3]
```diff
@@ -55,26 +55,26 @@ end
 
 """
     fftshift2d(mat::AbstractArray{T, N}) where {T, N}
-    Short-hand for fftshift(mat, (1,2)) performing only a 2D ft.
-    See fft for details.
+    Short-hand for fftshift(mat, (1,2)).
+    See fftshift for details.
```
Comment (roflmaostc) [inline src/fftshift_alternatives.jl]: `fftshift`

File: Project.toml [hunk 4]
```diff
@@ -6,6 +6,7 @@ version = "0.3.0"
 [deps]
 ChainRulesCore = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"
 FFTW = "7a1cc6ca-52ef-59f5-83cd-3a7055c09341"
+FastTransforms = "057dd010-8810-581a-b7be-e3fc3b93f78c"
```
Comment (roflmaostc) [inline Project.toml]: What is it needed for?

File: src/resampling.jl [hunk 5]
```diff
@@ -1,4 +1,5 @@
-export resample, resample_by_FFT, resample_by_RFFT, upsample2_abs2, upsample2, upsample2_1D
+export resample, resample_by_FFT, resample_by_RFFT, upsample2_abs2, upsample2, upsample2_1D, resample_var, resample_czt, barrel_pin
+using FastTransforms
```
Comment (roflmaostc) [inline src/resampling.jl]: if you need it, put it in FourierTools.jl file

File: src/resampling.jl [hunk 6]
```diff
@@ -129,3 +130,221 @@ Upsamples by a factor of two and applies the abs2 operation. The code is optimiz
...
+
+"""
+    resample_czt(arr, rel_zoom; shear=nothing, shear_dim=nothing, fix_nyquist=false, new_size = size(arr), rel_pad=0.2)
+
+resamples the image with fixed factors or a list of separable functions using the chirp z transform algorithm.
+The data is first padded by a relative amount `rel_pad` which is needed to avoid wrap-around problems.
+As opposed to `resample()`, this routine allows for arbitrary non-integer zoom factors.
+It is reasonably fast but only allows a stretch (via `rel_zoom`) and a shift (via `shear` in pixels) per line or column
+
+normalize=true by default multiplies by an appropriate factor so that the array size is included in the scaling.
```
Comment (roflmaostc) [inline src/resampling.jl]: `normalize=true`

File: src/resampling.jl [hunk 7]
```diff
@@ -129,3 +130,221 @@ Upsamples by a factor of two and applies the abs2 operation. The code is optimiz
...
+        my_zoom = 1.0
+        # case of a list of zoom numbers
+        if (isa(rel_zoom, Tuple) && isa(rel_zoom[1], Number)) || isa(rel_zoom, Number)
+            my_zoom = rel_zoom[d]
+            f_res = ift(arr, d)
+            myshear = 0
+            if !isnothing(shear)
+                myshear =  shear[d]
+            end
+            if myshear != 0
```
Comment (roflmaostc) [inline src/resampling.jl]: `if !iszero(myshear)`

File: src/resampling.jl [hunk 8]
```diff
@@ -129,3 +130,221 @@ Upsamples by a factor of two and applies the abs2 operation. The code is optimiz
...
+                # one-d since a slice was selected
+                f_res = ift(slice,1) 
+                if myshear != 0
+                    FourierTools.apply_shift_strength!(f_res, f_res, shifts[p], 1, 1, myshear, fix_nyquist)                
+                end
+                f_res = let 
+                    if T<:Real
+                        real(FourierTools.czt_1d(f_res, my_zoom, 1))
+                    else
+                        T.(FourierTools.czt_1d(f_res, my_zoom, 1))
```
Comment (roflmaostc) [inline src/resampling.jl]: That does a full allocation. Required? Why is it needed?

File: src/resampling.jl [hunk 9]
```diff
@@ -129,3 +130,221 @@ Upsamples by a factor of two and applies the abs2 operation. The code is optimiz
...
+            error("expected list of numbers or list of functions as argument `rel_zoom`")
+        end
+    end
+    res = select_region(arr, new_size=new_size)
+    return res
+end
+
+"""
+    barrel_pin(arr, rel=0.5)
+emulates a barrel (rel>0) or a pincushion (rel<0) distortion. The distortions are calculated using resample_czt() with separable quadratic zooms.
```
Comment (roflmaostc) [inline src/resampling.jl]: `rel>0`

File: src/resampling.jl [hunk 10]
```diff
@@ -129,3 +130,221 @@ Upsamples by a factor of two and applies the abs2 operation. The code is optimiz
...
+
+julia> c = barrel_pin(a,-0.5) # strong pin-cushion distortion
+
+julia> @ve a,b,c # visualize distortions
+```
+"""
+function barrel_pin(arr::AbstractArray{T,N}, rel=0.5) where {T,N}
+    RT = real(T)
+    fk = x -> one(RT) + RT(rel) .* (x-RT(0.5))^2
+    fkts = Tuple(fk for n in 1:ndims(arr))
```
Comment (roflmaostc) [inline src/resampling.jl]: `ntuple(n -> )` to make it stable

File: src/resampling.jl [hunk 11]
```diff
@@ -129,3 +130,221 @@ Upsamples by a factor of two and applies the abs2 operation. The code is optimiz
...
+    return resample_czt(arr, fkts)
+end
+
+"""
+    resample_var(img, rel_shift, dim=(2,))
+applies a variable shift `rel_shift` measured in pixel to each pixel in the source `img`.
```
Comment (roflmaostc) [inline src/resampling.jl]: new line in front of

**Comment by roflmaostc** [comment]
Apparently code coverage also dropped down. Can you increase that and check which lines are missing?

File: Project.toml (hunk 4 above)
Comment (RainerHeintzmann) [reply Project.toml]: nuFFT

File: src/resampling.jl (hunk 5 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: Are you sure this is the right position, as there seems no other export in FourierTools.jl

Comment (RainerHeintzmann) [reply src/resampling.jl]: I double checked. All other exports are also in the individual files.

File: src/resampling.jl (hunk 6 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: done

**Review by RainerHeintzmann — APPROVED** [review]

File: Project.toml (hunk 4 above)
Comment (roflmaostc) [reply Project.toml]: Did you compare with https://github.com/JuliaMath/NFFT.jl?

File: src/fftshift_alternatives.jl (hunk 3 above)
Comment (roflmaostc) [reply src/fftshift_alternatives.jl]: please fix formatting of fftshift

File: src/fftshift_alternatives.jl (hunk 2 above)
Comment (roflmaostc) [reply src/fftshift_alternatives.jl]: formatting

File: src/fftshift_alternatives.jl (hunk 1 above)
Comment (roflmaostc) [reply src/fftshift_alternatives.jl]: formatting

File: src/resampling.jl (hunk 5 above)
Comment (roflmaostc) [reply src/resampling.jl]: but not the using FastTransforms. That belongs in the FourierTools.jl file

File: src/resampling.jl (hunk 6 above)
Comment (roflmaostc) [reply src/resampling.jl]: I mean also the formatting in line 152. Please code related text in ` `

Comment (roflmaostc) [reply src/resampling.jl]: `

File: src/resampling.jl (hunk 7 above)
Comment (roflmaostc) [reply src/resampling.jl]: still todo

File: src/resampling.jl (hunk 8 above)
Comment (roflmaostc) [reply src/resampling.jl]: again

File: src/resampling.jl (hunk 7 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: I disagree. We have to test for the user argument not something that is anyway set to zero in the line before.

File: Project.toml (hunk 4 above)
Comment (RainerHeintzmann) [reply Project.toml]: Not really.

File: src/resampling.jl (hunk 6 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: I removed this comment as the function does not have a normalize argument.

File: src/resampling.jl (hunk 8 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: I don't know what you mean. I see 40 lines of code here. What exactly performs a "full" allocation?

File: src/resampling.jl (hunk 9 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: done

File: src/resampling.jl (hunk 10 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: OK

File: src/resampling.jl (hunk 11 above)
Comment (RainerHeintzmann) [reply src/resampling.jl]: done

**Review by RainerHeintzmann — APPROVED** [review]
Done most of the changes, except the memory thing and one that I disagree with.

File: src/resampling.jl (hunk 8 above)
Comment (roflmaostc) [reply src/resampling.jl]: `   T.(FourierTools.czt_1d(f_res, my_zoom, 1))` why is the `T.` needed?

File: src/resampling.jl (hunk 7 above)
Comment (roflmaostc) [reply src/resampling.jl]: Is just good style, always use `iszero`

File: src/fftshift_alternatives.jl (hunk 3 above)
Comment (RainerHeintzmann) [reply src/fftshift_alternatives.jl]: done

File: src/resampling.jl (hunk 7 above)
Comment (roflmaostc) [reply src/resampling.jl]: `! iszero(myshear)`

Comment (RainerHeintzmann) [reply src/resampling.jl]: Other line. Not I get it.
Done

Comment (RainerHeintzmann) [reply src/resampling.jl]: Now I get it is what I meant ;-)

**Review by RainerHeintzmann — APPROVED** [review]

**Comment by roflmaostc** [comment]
Can you create your own Pull Request again?
Tests still fail locally

**Pull request closed without merging**